File: cpp/n345_ReverseVowels.cpp
// https://leetcode-cn.com/problems/reverse-vowels-of-a-string/
#include <iostream>
#include <vector>
#include <sstream>
#include <string>
#include <unordered_map>
#include <algorithm>
using namespace std;

class Solution {
public:
    string reverseVowels(string s) {
        string vowels = "aeiouAEIOU";
        int head = 0;
        int tail = s.size() - 1;
        while (head < tail){
            while ((vowels.find(s[head]) == vowels.npos) && (head < tail))
                head += 1;
            while ((vowels.find(s[tail]) == vowels.npos) && (head < tail))
                tail -= 1;
            if (head < tail){
                swap(s[head], s[tail]);
                head += 1;
                tail -= 1;
            }
        }
        return s;
    }
};

int main(){
    string s = "lEetcOde";
    Solution solu;
    cout << solu.reverseVowels(s) << endl;
}

File: python/n345_ReverseVowels.py
# https://leetcode-cn.com/problems/reverse-vowels-of-a-string/

class Solution:
    def reverseVowels(self, s: str) -> str:
        vowels = "aeiouAEIOU"
        head = 0
        tail = len(s) - 1
        while head < tail:
            if s[head] in vowels and s[tail] in vowels:
                s = s[:head] + s[tail] + s[head+1:tail] + s[head] + s[tail+1:]
                head += 1
                tail -= 1
            elif s[head] in vowels and s[tail] not in vowels:
                tail -= 1
            elif s[head] not in vowels and s[tail] in vowels:
                head += 1
            else:
                head += 1
                tail -= 1 
        return s                
            
            
if __name__ == "__main__":
    solu = Solution()
    s = "leEtcOde"
    print(solu.reverseVowels(s))